src/Array.cpp

```diff
@@ -57,27 +57,44 @@ int64 Array::length() const {
 }
 
 LocalReference<jarray> Array::getNew(Type elementType, jclass elementClass, jsize size) {
-   Env env;
+    Env env;
+
+    switch (elementType) {
+        case Type::Boolean: return env.newBooleanArray(size).as<jarray>();
+        case Type::Byte: return env.newByteArray(size).as<jarray>();
+        case Type::Char: return env.newCharArray(size).as<jarray>();
+        case Type::Short: return env.newShortArray(size).as<jarray>();
+        case Type::Int: return env.newIntArray(size).as<jarray>();
+        case Type::Long: return env.newLongArray(size).as<jarray>();
+        case Type::Float: return env.newFloatArray(size).as<jarray>();
+        case Type::Double: return env.newDoubleArray(size).as<jarray>();
+        case Type::Reference:
+        default:
+            assert(elementType == Type::Reference);
+            return env.newObjectArray(size, elementClass).as<jarray>();
+    }
+}
+
+BinaryNode* Array::getBinary(Env& env, jarray array) {
+    SimpleRefHolder<BinaryNode> rv(new BinaryNode);
+
+    jsize size = env.getArrayLength(array);
+    rv->preallocate(size);
 
-   switch (elementType) {
-      case Type::Boolean: return env.newBooleanArray(size).as<jarray>();
-      case Type::Byte: return env.newByteArray(size).as<jarray>();
-      case Type::Char: return env.newCharArray(size).as<jarray>();
-      case Type::Short: return env.newShortArray(size).as<jarray>();
-      case Type::Int: return env.newIntArray(size).as<jarray>();
-      case Type::Long: return env.newLongArray(size).as<jarray>();
-      case Type::Float: return env.newFloatArray(size).as<jarray>();
-      case Type::Double: return env.newDoubleArray(size).as<jarray>();
-      case Type::Reference:
-      default:
-         assert(elementType == Type::Reference);
-         return env.newObjectArray(size, elementClass).as<jarray>();
-   }
-}
-
-QoreListNode* Array::getList(Env& env, jarray array, jclass arrayClass) {
+    for (jsize i = 0; i < size; ++i) {
+        jbyte byte = env.getByteArrayElement(static_cast<jbyteArray>(array), i);
+        reinterpret_cast<jbyte*>(const_cast<void*>(rv->getPtr()))[i] = byte;
```

Comment:
> this is misusning the BinaryNode API, getPtr returns read only pointer to inner data, so either introduce a new one, that would return a non-const variant (which I don't like, see std::string for example, there is no such thing, because you don't want to let outsiders to manipulate your inner structures), or use just append, because append shouldn't reallocate the binaryNode, if there is enough space left (preallocated). I look into append and it will reallocate always, which is probably wrong.

> Also it seems wrong to me casting BinaryNode element to (jbyte*).

> Btw why BinrayNode inner implementation is using void* if it anyway considers it as char* for every pointer operation? I would say, that BinaryNode should use char* as internal representation, byte can be understood as char in C++, and then jbyte should be converted to char, resp jbyte* to char*.

This feedback is good, and I'm happy for all of these issues to be fixed - let's talk about how it can be done.

> also based on https://stackoverflow.com/questions/8439233/how-to-convert-jbytearray-to-native-char-in-jni calling getByteArrayElement should be followed with ReleaseByteArrayElements...

In fact `Env::getByteArrayElement()` (note the lack of an `s` on the end) does not use the JNI function `GetByteArrayElements()`, which requires a release but rather `GetBooleanArrayRegion()`, which does not.

Reply:
@sejvlond another problem is with `ReferenceHolder<>` which takes an `ExceptionSink*` arg in the constructor, which is often a local var (particularly in the jni module which uses C++ and Java exceptions instead), so creating a move constructor could cause more problems than it would solve - basically we need to migrate away from `ExceptionSink` and just use C++ exceptions, but let's please discuss what to do in the short term

Reply:
updated; please review the new changes

Reply:
I don't like the new solution either :) but okay, let's say it's a temporary one and at least the return value is clear that must be dereferenced via the ReferenceHolder. I would really like to change the exceptions and these return values.